### PR TITLE
Add a CommTag class with a stable hash

### PR DESCRIPTION
**Pull request by matthiasdiener**
TODO:

- [x] add tests

### CODE REVIEW

**Comment by matthiasdiener** [comment]
This is still a draft, but it's ready for a first look @inducer @majosm

File: grudge/trace_pair.py [hunk 1]
```diff
@@ -318,8 +319,20 @@ def interior_trace_pair(dcoll: DiscretizationCollection, vec) -> TracePair:
...
+    def __hash__(self):
+        return hash(tuple(str(type(self)).encode("ascii")))
+
+    def __eq__(self, other):
+        return isinstance(other, type(self))
+
+
 def interior_trace_pairs(dcoll: DiscretizationCollection, vec, *,
-        comm_tag: Hashable = None, tag: Hashable = None,
+        comm_tag: CommTag = None, tag: Hashable = None,
```
Comment (inducer) [inline grudge/trace_pair.py:335]: I don't agree with this change. There's  no good reason why other `Hashable`s should not be accepted.

File: grudge/trace_pair.py [hunk 2]
```diff
@@ -318,8 +319,20 @@ def interior_trace_pair(dcoll: DiscretizationCollection, vec) -> TracePair:
     return local_interior_trace_pair(dcoll, vec)
 
 
+@dataclass(frozen=True)  # for KeyBuilder support
+class CommTag:
```
Comment (inducer) [inline grudge/trace_pair.py:323]: It doesn't really make sense for all `CommTag`s to be subclasses of this. Some may have data contained in them.

File: grudge/trace_pair.py [hunk 3]
```diff
@@ -318,8 +319,20 @@ def interior_trace_pair(dcoll: DiscretizationCollection, vec) -> TracePair:
...
+    """A communication tag with a hash value that is stable across
+    runs, even without setting ``PYTHONHASHSEED``."""
+    @memoize_method
+    def __hash__(self):
+        return hash(tuple(str(type(self)).encode("ascii")))
+
+    def __eq__(self, other):
+        return isinstance(other, type(self))
```
Comment (inducer) [inline grudge/trace_pair.py:331]: This is incorrect: It accepts arbitrary subclasses for `other`, and violates the property of "`__eq__` implies hash equality."

File: grudge/trace_pair.py (hunk 1 above)
Comment (matthiasdiener) [reply grudge/trace_pair.py]: I undid this change in [9b2ced4](https://github.com/inducer/grudge/pull/319/commits/9b2ced47c82129bfdc381292cca900ba0da2d240). I left the warning in https://github.com/inducer/grudge/pull/319/files#diff-5fa6e6b713d37028c16f1bbc5a6d0b4547ecba0f86c8a52eb30e1ba1dc2614e4R518 in  place for now, would you like me to remove it?

File: grudge/trace_pair.py (hunk 2 above)
Comment (matthiasdiener) [reply grudge/trace_pair.py:323]: I added a small test in [9a63df4](https://github.com/inducer/grudge/pull/319/commits/9a63df4e774e517918ae944dbc89e8b2bac17568) with subclassed dataclasses, is that what you had in mind?

File: grudge/trace_pair.py (hunk 3 above)
Comment (matthiasdiener) [reply grudge/trace_pair.py]: This is (hopefully) fixed in [6918c63](https://github.com/inducer/grudge/pull/319/commits/6918c63ca6a69829ed9bfb805045657020a2187e)

File: grudge/trace_pair.py (hunk 2 above)
Comment (matthiasdiener) [reply grudge/trace_pair.py:322]: Ok, would you like to keep the `CommTag` class as an optional way to construct the comm tags? i removed the warning in 945406730bd968fdde9688dbb7ce604a329b7110.

**Comment by matthiasdiener** [comment]
It's unlikely this will be needed for deterministic execution, closing.

**Pull request closed without merging**